src/clients/FlaunchZapClient.ts

```diff
@@ -0,0 +1,168 @@
+  flaunchAt?: bigint;
+  premineAmount?: bigint;
+}
+
+export interface FlaunchWithRevenueManagerParams extends FlaunchParams {
+  revenueManagerInstanceAddress: Address;
+}
+
+export interface FlaunchWithRevenueManagerIPFSParams
+  extends Omit<FlaunchWithRevenueManagerParams, "tokenUri">,
+    IPFSParams {}
+
+/**
+ * Extended client for interacting with the FlaunchZap contract with write capabilities
+ */
+export class ReadWriteFlaunchZap extends ReadFlaunchZap {
+  chainId: number;
+  declare contract: ReadWriteContract<FlaunchZapABI>;
+
+  constructor(
+    chainId: number,
+    address: Address,
+    drift: Drift<ReadWriteAdapter> = createDrift()
+  ) {
+    super(address, drift);
+    this.chainId = chainId;
+  }
+
+  /**
+   * Creates a new flaunch with revenue manager configuration
+   * @param params - Parameters for the flaunch with revenue manager
+   * @param params.name - The name of the token
+   * @param params.symbol - The symbol of the token
+   * @param params.tokenUri - The URI containing the token metadata
+   * @param params.fairLaunchPercent - Percentage of total supply to be used in fair launch (0-100)
+   * @param params.fairLaunchDuration - Duration of fair launch in seconds
+   * @param params.initialMarketCapUSD - Initial market cap in USD
+   * @param params.creator - Address of the token creator
+   * @param params.creatorFeeAllocationPercent - Percentage of fees allocated to creator (0-100)
+   * @param params.protocolRecipient - Address to receive protocol fees
+   * @param params.protocolFeePercent - Percentage of fees allocated to protocol (0-100)
+   * @param params.flaunchAt - Optional timestamp when the flaunch should start
+   * @param params.premineAmount - Optional amount of tokens to premine
+   * @returns Transaction response for the flaunch creation
+   */
+  flaunchWithRevenueManager(params: FlaunchWithRevenueManagerParams) {
+    const initialMCapInUSDCWei = parseUnits(
+      params.initialMarketCapUSD.toString(),
+      6
+    );
+    const initialPriceParams = encodeAbiParameters(
+      [
+        {
+          type: "uint256",
+        },
+      ],
+      [initialMCapInUSDCWei]
+    );
+
+    const fairLaunchInBps = BigInt(params.fairLaunchPercent * 100);
+    const creatorFeeAllocationInBps = params.creatorFeeAllocationPercent * 100;
+
+    return this.contract.write("flaunch", {
+      _flaunchParams: {
+        name: params.name,
+        symbol: params.symbol,
+        tokenUri: params.tokenUri,
+        initialTokenFairLaunch: (this.TOTAL_SUPPLY * fairLaunchInBps) / 10_000n,
+        fairLaunchDuration: BigInt(params.fairLaunchDuration),
+        premineAmount: params.premineAmount,
+        creator: params.creator,
+        creatorFeeAllocation: creatorFeeAllocationInBps,
+        flaunchAt: params.flaunchAt,
+        initialPriceParams,
+        feeCalculatorParams: "0x",
+      },
+      _treasuryManagerParams: {
+        manager: params.revenueManagerInstanceAddress,
```

Comment:
This will be fine as long as this is an existing revenue manager address. If it was the implementation then we would need `initializeData`

Reply:
for protocols deploying via the revenue manager, they have to pass in their instance address as the protocolReceiver is tied to each of their individual managers. The implementation address wouldn't be passed here

Reply:
Perfect, figured that was the case 🙌